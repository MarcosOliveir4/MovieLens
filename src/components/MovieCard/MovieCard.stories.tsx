import { Meta, StoryObj } from '@storybook/react'
import MovieCard from '.'

export default {
  title: 'MovieCard',
  component: MovieCard
} as Meta

export const Default: StoryObj = {}
