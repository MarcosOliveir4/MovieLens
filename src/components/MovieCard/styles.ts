import styled from 'styled-components'

export const Wrapper = styled.div`
  width: 340px;
  height: 620px;

  margin-bottom: 2.5rem;

  gap: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border-radius: 4px;

  padding: 1rem;
  background-color: ${({ theme }) => theme.colors.softBlack};
`

export const WrapperImg = styled.div`
  position: relative;
  height: 500px;
  border-radius: 4px;
  overflow: hidden;
`

export const Title = styled.h2``

export const VoteAverage = styled.p`
  gap: 0.5rem;
  display: flex;

  svg {
    color: ${({ theme }) => theme.colors.yellow};
  }
`

export const DetailsButton = styled.button`
  background-color: ${({ theme }) => theme.colors.yellow};
  border: 2px solid ${({ theme }) => theme.colors.yellow};
  border-radius: 4px;
  padding: 1rem 0.5rem;
  font-weight: bold;

  transition: all 0.5s;
  a {
    color: ${({ theme }) => theme.colors.black};
    text-decoration: none;
  }

  cursor: pointer;
`
