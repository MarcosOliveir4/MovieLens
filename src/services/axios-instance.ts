'use client'
import axios from 'axios'

export const api = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL,
  headers: {
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_API_KEY}`
  }
})

api.interceptors.response.use(
  (response) => {
    return response
  },
  (error) => {
    return Promise.reject(error)
  }
)
