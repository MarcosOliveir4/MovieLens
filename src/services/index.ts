export * from './tmdb'
export * from './tmdb.models'
