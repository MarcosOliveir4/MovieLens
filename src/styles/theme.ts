const Theme = {
  background: '#000',
  colors: {
    warning: '#f7d354',
    darkWarning: '#b8930c',

    fontColor: '#fff'
  }
}

export default Theme
