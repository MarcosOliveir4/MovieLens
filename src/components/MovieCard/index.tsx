import { FaStar } from 'react-icons/fa'

import * as S from './styles'
import { Movie } from '@/services'
import Image from 'next/image'
import { useRouter } from 'next/navigation'

interface MovieCardInterface {
  movie: Movie
}

const imageUrl = process.env.NEXT_PUBLIC_API_URL_IMG

const MovieCard = ({ movie }: MovieCardInterface) => {
  const router = useRouter()

  return (
    <S.Wrapper data-testid="movie-card">
      <S.WrapperImg>
        <Image
          data-testid="movie-card-img"
          src={imageUrl + movie.poster_path}
          alt={movie.title}
          fill
          priority
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          style={{ objectFit: 'cover' }}
          overrideSrc={imageUrl + movie.poster_path}
        />
      </S.WrapperImg>
      <S.Title data-testid="movie-card-title">{movie.title}</S.Title>
      <S.VoteAverage data-testid="movie-card-vote-average">
        <FaStar data-testid="movie-card-vote-average-icon" />{' '}
        {movie.vote_average}
      </S.VoteAverage>
      <S.DetailsButton
        data-testid="movie-card-details-button"
        onClick={() => router.push(`/movie/${movie.id}`)}
      >
        Detalhes
      </S.DetailsButton>
    </S.Wrapper>
  )
}
export default MovieCard
