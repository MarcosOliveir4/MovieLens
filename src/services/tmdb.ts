import { api } from './axios-instance'
import { TopRatedMovies } from '@/services/tmdb.models'

const getTopRatedMovies = async () => {
  try {
    const { data } = await api.get<TopRatedMovies>('top_rated')
    return data
  } catch (error) {
    return Promise.reject(error)
  }
}

export { getTopRatedMovies }
