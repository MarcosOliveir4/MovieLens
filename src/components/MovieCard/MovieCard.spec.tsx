import { render, screen } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import userEvent from '@testing-library/user-event'

import MovieCard from '.'
import { Movie } from '@/services'
import theme from '@/styles/theme'

const movieMock: Movie = {
  adult: false,
  backdrop_path: '/m4TUa2ciEWSlk37rOsjiSIvZDXE.jpg',
  genre_ids: [16, 10751, 14],
  id: 129,
  original_language: 'ja',
  original_title: '千と千尋の神隠し',
  overview:
    'A young girl, Chihiro, becomes trapped in a strange new world of spirits. When her parents undergo a mysterious transformation, she must call upon the courage she never knew she had to free her family.',
  popularity: 103.603,
  poster_path: '/39wmItIWsg5sZMyRUHLkWBcuVCM.jpg',
  release_date: '2001-07-20',
  title: 'Spirited Away',
  video: false,
  vote_average: 8.537,
  vote_count: 16633
}

const imageUrl = process.env.NEXT_PUBLIC_API_URL_IMG

const mockPush = jest.fn()

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      prefetch: () => null,
      push: mockPush
    }
  }
}))

describe('<MovieCard />', () => {
  beforeEach(() => {
    render(
      <ThemeProvider theme={theme}>
        <MovieCard movie={movieMock} />
      </ThemeProvider>
    )
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render the component', async () => {
    const container = screen.getByTestId('movie-card')
    expect(container).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should render img correct', () => {
    const img = screen.getByTestId('movie-card-img')
    expect(img).toHaveAttribute('src', imageUrl + movieMock.poster_path)
    expect(img).toHaveAttribute('alt', movieMock.title)
  })

  it('should render title correct', () => {
    const title = screen.getByTestId('movie-card-title')
    expect(title).toHaveTextContent(movieMock.title)
  })

  it('should render vote average correct', () => {
    const voteAverage = screen.getByTestId('movie-card-vote-average')
    const voteAverageIcon = screen.getByTestId('movie-card-vote-average-icon')

    expect(voteAverage).toHaveTextContent(movieMock.vote_average.toString())
    expect(voteAverageIcon).toBeInTheDocument()
  })

  it('should render and emitter action details button correct', async () => {
    const detailsButton = screen.getByTestId('movie-card-details-button')
    expect(detailsButton).toBeInTheDocument()

    await userEvent.click(detailsButton)
    expect(mockPush).toHaveBeenCalledTimes(1)
    expect(mockPush).toHaveBeenCalledWith(`/movie/${movieMock.id}`)
  })
})
