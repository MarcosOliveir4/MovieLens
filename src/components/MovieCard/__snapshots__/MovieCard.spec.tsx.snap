// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> should render the component 1`] = `
.c0 {
  width: 340px;
  height: 620px;
  margin-bottom: 2.5rem;
  gap: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border-radius: 4px;
  padding: 1rem;
  background-color: #121212;
}

.c1 {
  position: relative;
  height: 500px;
  border-radius: 4px;
  overflow: hidden;
}

.c2 {
  gap: 0.5rem;
  display: flex;
}

.c2 svg {
  color: #f7d354;
}

.c3 {
  background-color: #f7d354;
  border: 2px solid #f7d354;
  border-radius: 4px;
  padding: 1rem 0.5rem;
  font-weight: bold;
  transition: all 0.5s;
  cursor: pointer;
}

.c3 a {
  color: #000;
  text-decoration: none;
}

<div
  class="c0"
  data-testid="movie-card"
>
  <div
    class="c1"
  >
    <img
      alt="Spirited Away"
      data-nimg="fill"
      data-testid="movie-card-img"
      decoding="async"
      sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
      src="https://image.tmdb.org/t/p/w500/39wmItIWsg5sZMyRUHLkWBcuVCM.jpg"
      srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Fw500%2F39wmItIWsg5sZMyRUHLkWBcuVCM.jpg&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
    />
  </div>
  <h2
    class=""
    data-testid="movie-card-title"
  >
    Spirited Away
  </h2>
  <p
    class="c2"
    data-testid="movie-card-vote-average"
  >
    <svg
      data-testid="movie-card-vote-average-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      />
    </svg>
     
    8.537
  </p>
  <button
    class="c3"
    data-testid="movie-card-details-button"
  >
    Detalhes
  </button>
</div>
`;
